ui: expose system jobs

benesch opened the pull request:
Add a new page to the admin UI that provides a graphical, interactive
interface to `SHOW JOBS`.

![screenshot 2017-08-02 18 36 35](https://user-images.githubusercontent.com/882976/28898043-9cf373ee-77b1-11e7-890b-e87c91a75688.png)

Closes #13604.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17253)
<!-- Reviewable:end -->

jordanlewis commented:
Nice!! I'll leave the UI review to the experts

---

Reviewed 21 of 21 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 1019 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3K0mHAUMS4ZqYjFcy:-Kq3K0mHAUMS4ZqYjFcz:bl8hu5j) ([raw file](https://github.com/cockroachdb/cockroach/blob/fba0dd2edbec2a9468903b08ef5dfa174253d964/pkg/server/admin.go#L1019)):*
> ```Go
> 			SELECT id, type, description, username, descriptor_ids, status,
> 				created, started, finished, modified, fraction_completed, error
> 			FROM [SHOW JOBS]
> ```

Why not FROM crdb_internal.jobs? I'm not positive but it seems likely that it's more efficient to do it that way.

---

*[pkg/server/admin.go, line 1257 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3KblhsLl-CSKEgJ9k:-Kq3KblhsLl-CSKEgJ9l:blfdcad) ([raw file](https://github.com/cockroachdb/cockroach/blob/fba0dd2edbec2a9468903b08ef5dfa174253d964/pkg/server/admin.go#L1257)):*
> ```Go
> 	if dst == nil {
> 		return errors.Errorf("nil destination pointer passed in")
> 	}
> ```

nice DRYing

---

*[pkg/server/admin.go, line 1308 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3KSEQ0y0NJjPZgqyF:-Kq3KSEQ0y0NJjPZgqyG:bg40s89) ([raw file](https://github.com/cockroachdb/cockroach/blob/fba0dd2edbec2a9468903b08ef5dfa174253d964/pkg/server/admin.go#L1308)):*
> ```Go
> 		*d = s.Time
> 
> 	case **time.Time:
> ```

What's this for? Maybe add a comment to explain why its a ** unlike the rest of the cases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-Kq3JvPjxaq1iS-D45ey:b-j8jour)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 13 of 21 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/server/admin.go, line 1019 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3K0mHAUMS4ZqYjFcy:-Kq3LZ1F3SGxRp_JnE0R:bik8h5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/fba0dd2edbec2a9468903b08ef5dfa174253d964/pkg/server/admin.go#L1019)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Why not FROM crdb_internal.jobs? I'm not positive but it seems likely that it's more efficient to do it that way.
</blockquote></details>

Just preferring public API to private. AIUI, @knz did a refactor so that `SHOW FOO` queries are identical to querying the underlying vtable via [`p.delegateQuery`](https://github.com/benesch/cockroach/blob/jobs-ui/pkg/sql/show.go#L572). So once we push filters down to the vtable constructors we should see that speedup in `SHOW JOBS` as well.

---

*[pkg/server/admin.go, line 1257 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3KblhsLl-CSKEgJ9k:-Kq3McXD9nwLWvRZexNw:b32vin) ([raw file](https://github.com/cockroachdb/cockroach/blob/fba0dd2edbec2a9468903b08ef5dfa174253d964/pkg/server/admin.go#L1257)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nice DRYing
</blockquote></details>

:D

---

*[pkg/server/admin.go, line 1308 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3KSEQ0y0NJjPZgqyF:-Kq3MbgLL5AyM8Lj1KP-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fba0dd2edbec2a9468903b08ef5dfa174253d964/pkg/server/admin.go#L1308)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What's this for? Maybe add a comment to explain why its a ** unlike the rest of the cases.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253)*
<!-- Sent from Reviewable.io -->

danhhz commented:
yolo'd this together in an evening, eh? looks great to me, but :sciencedog:

---

Review status: 13 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 1308 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3KSEQ0y0NJjPZgqyF:-Kq3V_yWWZ2rqGaqrFfA:b3926w4) ([raw file](https://github.com/cockroachdb/cockroach/blob/fba0dd2edbec2a9468903b08ef5dfa174253d964/pkg/server/admin.go#L1308)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Any reason you can't use the zero value to mean unset and collapse the two cases?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-Kq3VhYhCD9kN-Xcpncm:b-slfvqe)*
<!-- Sent from Reviewable.io -->

dianasaur323 commented:
Awesome! LGTM. Only question from me is why we need that "Duration: 00:00:00" under completed backups. Is that just dummy data that will be filled in once the backups are actually run?

---

Review status: 13 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-Kq3Yb_j-DzpGRXIJczD:baxg6em)*
<!-- Sent from Reviewable.io -->

dianasaur323 commented:
@kuanluo could you take a look at these pages? Looks good to me, but that's your wheelhouse.

dianasaur323 commented:
I'm sorry - just 2 more clarifying questions from me. It looks like schema changes will also be included since this is all coming out of the jobs table, correct? Finally, I think we had a discussion a while ago about pagination. It looks like you've implemented filtering, in which case is there some sort of "maximum time" in which I won't be able to see super old jobs? Should we surface that limitation in the title?

---

Review status: 13 of 21 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-Kq3_2-z-KMILZ3U1W0S:b-cdqp20)*
<!-- Sent from Reviewable.io -->

couchand commented:
Reviewed 3 of 21 files at r1, 6 of 8 files at r2.
Review status: 16 of 21 files reviewed at latest revision, 20 unresolved discussions.

---

*[pkg/server/admin.go, line 1340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3VgffvWBxJmHOwRpQ:-Kq3VgffvWBxJmHOwRpR:bh0vgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/server/admin.go#L1340)):*
> ```Go
> 	if len(row) != len(dsts) {
> 		return fmt.Errorf(
> 			"ScanAll: row has %d columns but only %d dests provided", len(row), len(dsts))
> ```

nit: "only" will seem odd if too many dests are provided

---

*[pkg/ui/src/views/jobs/index.tsx, line 15 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3YnKYa391cxjD1jPu:-Kq3YnKYa391cxjD1jPv:bo56xtl) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L15)):*
> ```XML
> import { SortSetting } from "src/views/shared/components/sortabletable";
> import { SortedTable } from "src/views/shared/components/sortedtable";
> import { TimestampToMoment } from "src/util/convert";
> ```

We're trying to enforce more consistency on imports now.  The general goal is:

```
<external imports, sorted alphabetically>
<blank line>
<internal imports, sorted alphabetically>
```

---

*[pkg/ui/src/views/jobs/index.tsx, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3drAvEKHZb7vyqGFQ:-Kq3drAvEKHZb7vyqGFR:b-kbmsr2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L75)):*
> ```XML
>       return "Duration: " + formatDuration(moment.duration(finished.diff(started)));
>     }
>     return formatDuration(moment.duration(modified.diff(started) / this.props.job.fraction_completed)) +
> ```

We might not want to guess this, and instead just report time elapsed.  It will inevitably be unstable, and we've all been frustrated by progress bars with wildly inaccurate "time remaining" estimates.

---

*[pkg/ui/src/views/jobs/index.tsx, line 89 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3eS342qGM_eq5VQLe:-Kq3eS342qGM_eq5VQLf:bcas76v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L89)):*
> ```XML
>             className="jobs-table__progress-bar"
>           />}
>         <span title={percent.toFixed(5) + "%"}>{percent.toFixed(1) + "%"}</span>
> ```

Five digits seems like a ridiculous precision... is that likely to be useful?

---

*[pkg/ui/src/views/jobs/index.tsx, line 91 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ee-XN6xmKDhsZacu:-Kq3ee-XN6xmKDhsZacv:bbdb9zr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L91)):*
> ```XML
>         <span title={percent.toFixed(5) + "%"}>{percent.toFixed(1) + "%"}</span>
>         <br />
>         <em>{this.renderDuration()}</em>
> ```

This might be better done with explicit styling.

---

*[pkg/ui/src/views/jobs/index.tsx, line 99 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ZlQLNfMApYRRz7dt:-Kq3ZlQLNfMApYRRz7du:b-ql79rz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L99)):*
> ```XML
> //   https://github.com/Microsoft/TypeScript/issues/3960
> //
> // The variable name must start with a capital letter or TSX will not recognize
> ```

nit: it's JSX that expects it to be capitalized (I suppose you could say TSX does as well, but only because it inherits the requirement).  AFAIC this comment is unnecessary, perhaps others disagree.

---

*[pkg/ui/src/views/jobs/index.tsx, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3_NRNKs6lCeICa7Ig:-Kq3_NROrjONFZNs-PUD:bxaeeza) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L105)):*
> ```XML
> 
> interface JobsTableProps {
>   sort: SortSetting;
> ```

Making this table sortable in combinations with the limit option is pretty awkward.  Usually the expectation of a sort on a table showing limited results is that it will do the same thing as the corresponding SQL query, with the ORDER BY applied _before_ the LIMIT.  Since the API endpoint doesn't accept a sort parameter, this won't work as expected.  I'd suggest either exposing the sort as an additional API parameter, or removing the sort option entirely.

---

*[pkg/ui/src/views/jobs/index.tsx, line 118 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3f9BO0GfWRlgDxV4C:-Kq3f9BPLuj4gMLIpcmS:b-fm2243) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L118)):*
> ```XML
> 
> class JobsTable extends React.Component<JobsTableProps, {}> {
>   static title() {
> ```

What's this for?

---

*[pkg/ui/src/views/jobs/index.tsx, line 146 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ftzOqeNvWIFVMH54:-Kq3ftzOqeNvWIFVMH55:br60nx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L146)):*
> ```XML
>             options={statusOptions}
>             selected={this.props.status}
>             onChange={(selected: DropdownOption) => this.props.setStatus(selected.value)}
> ```

Inlining handlers is not ideal since it prevents a component from shallowly checking its props for changes.  It's better to make a function elsewhere (usually as a method of this component) that will be identical across rerenders.

---

*[pkg/ui/src/views/jobs/index.tsx, line 147 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ffy4gBjYlxEtNHXI:-Kq3ffy4gBjYlxEtNHXJ:beyrkrb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L147)):*
> ```XML
>             selected={this.props.status}
>             onChange={(selected: DropdownOption) => this.props.setStatus(selected.value)}
>             />
> ```

I think the ends of these tags should line up with their start (and we should have a lint rule for it...)

---

*[pkg/ui/src/views/jobs/index.tsx, line 174 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3gjTZQuPU9jumutoY:-Kq3gjTZQuPU9jumutoZ:b-iuegz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L174)):*
> ```XML
>             className="jobs-table"
>             rowClass={job => "jobs-table__row--" + job.status}
>             columns={[
> ```

This column configuration should be defined above to avoid allocations and so that it's identical across rerenders.

---

*[pkg/ui/src/views/jobs/index.tsx, line 188 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3hIveBTdoFtxiilsW:-Kq3hIveBTdoFtxiilsX:b-cnr7zp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L188)):*
> ```XML
>               {
>                 title: "Start Time",
>                 cell: job => TimestampToMoment(job.started).fromNow(),
> ```

We should try to minimize calls to the `moment` constructor.  It's fairly expensive, and we're just repeating work if we do it on every render.  It's much better to do that once when we get the API results than repeatedly in component renders.  Moving it there might be a significant change, since the `timeoutFetch` and `CachedDataReducer` abstract away all of that side of things.  But we can still improve somewhat by moving it to the `mapStateToProps` function and only do it once per state change.

---

*[pkg/ui/src/views/jobs/index.tsx, line 211 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3gw3PQsPdZLPDiU6C:-Kq3gw3PQsPdZLPDiU6D:b-5k2pph) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L211)):*
> ```XML
> 
> export default connect(
>   (state: AdminUIState) => {
> ```

I prefer to extract the `mapStateToProps` and `mapDispatchToProps` (or in this case, just `actions`) so that the `connect` call can be one-lined.

---

*[pkg/ui/styl/base/palette.styl, line 39 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ierGzLmVFh3Bk4Dr:-Kq3ierGzLmVFh3Bk4Ds:bh1oei) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/base/palette.styl#L39)):*
> ```Stylus
> $notification-warning-fill-color = #FDF8E9
> $notification-warning-border-color = #F9E6AF
> $table-alert-fill-color = #fdefef
> ```

Looks like we're using upper-case hex codes.

---

*[pkg/ui/styl/pages/jobs.styl, line 5 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3i_mbw56pccne1T2Q:-Kq3i_mbw56pccne1T2R:b-fm2243) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/pages/jobs.styl#L5)):*
> ```Stylus
>     &--failed
>       & + &
>         border-top 1px solid $table-border-color
> ```

What's this for?

---

*[pkg/ui/styl/pages/jobs.styl, line 13 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3iwwmuXzmUmAy2SDl:-Kq3iwwnbgt-loWCFDog:b7so02y) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/pages/jobs.styl#L13)):*
> ```Stylus
>   &__cell
>     &--description
>       font-family Inconsolata-Regular
> ```

This should have a fallback generic family.

---

*[pkg/ui/styl/pages/jobs.styl, line 14 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ikwUe93Wc1NibJlX:-Kq3ikwVd5QV8Cp3D1Oa:b-415yq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/pages/jobs.styl#L14)):*
> ```Stylus
>     &--description
>       font-family Inconsolata-Regular
>       
> ```

Trailing whitespace.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253)*
<!-- Sent from Reviewable.io -->

kuanluo commented:
Looks great overall! Noticed a few different approaches comparing to the mock, flagging: 

1. Are we still going to show all queries in the last 7 days on one single scroll? If so, we might not need the "Show" button, or the rest.
2. For succeed queries, can we gray them out? I'm fine with leaving them dark, if we won't have multiple running queries on top of succeeded ones.
3. How are these queries ordered? The column title need to called out. 
4. In my mock, I have "success" replace "100%" when the query is finished, so I combined the column "progress" and "status." 

![image](https://user-images.githubusercontent.com/1486483/28679723-1ac16afa-72c2-11e7-87ee-895fa6c8418b.png)

In addition: are we going to have the ability to copy SQL to pause/cancel queries?
![image](https://user-images.githubusercontent.com/1486483/28679866-87fad458-72c2-11e7-9b38-b2889fa25c67.png)

danhhz commented:
Review status: 16 of 21 files reviewed at latest revision, 20 unresolved discussions, some commit checks pending.

---

*[pkg/ui/src/views/jobs/index.tsx, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3drAvEKHZb7vyqGFQ:-Kq3mxpbGspSRqn5Tc0n:bpsfsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L75)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

We might not want to guess this, and instead just report time elapsed.  It will inevitably be unstable, and we've all been frustrated by progress bars with wildly inaccurate "time remaining" estimates.
</blockquote></details>

anecdotally, we've been doing this manually in a sql shell for months and (at least for backup/restore, dunno about schema change) it's reasonably accurate. and having it has definitely been better than not having it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253)*
<!-- Sent from Reviewable.io -->

benesch commented:
@dianasaur, the 00:00:00 is because that backup completed literally instantly. It would be more useful on a longer backup.

@kuanluo, we don't support pause and cancel yet, which is why it's missing. :(

Thanks for all the reviews so far, everyone! Just a quick heads up that I'm swamped with other features for the 1.1 code freeze, so the more we can punt to August the better. I'm not sure how these things usually work, but in particular would love to land without full product/design QA: having this draft version merged will be hugely helpful as we debug backup/restore, and I'll have plenty of time for UX polishing in August.

@couchand, I responded to some of your points; anything I didn't respond to is just something I intend to fix.

---

Review status: 16 of 21 files reviewed at latest revision, 20 unresolved discussions, some commit checks pending.

---

*[pkg/server/admin.go, line 1308 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3KSEQ0y0NJjPZgqyF:-Kq3k0eqWvSOOrmclHlF:b-pf8f9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/fba0dd2edbec2a9468903b08ef5dfa174253d964/pkg/server/admin.go#L1308)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Any reason you can't use the zero value to mean unset and collapse the two cases?
</blockquote></details>

Only that JS will happily distinguish null from the empty time and polluting it with Go's empty-value weirdness makes me itchy.

---

*[pkg/ui/src/views/jobs/index.tsx, line 15 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3YnKYa391cxjD1jPu:-Kq3khmQg4nAOMWJ-Wdl:b26691j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L15)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

We're trying to enforce more consistency on imports now.  The general goal is:

```
<external imports, sorted alphabetically>
<blank line>
<internal imports, sorted alphabetically>
```
</blockquote></details>

Damn, I spent like 10m trying to make sense of the ordering rules last night. SGTM.

---

*[pkg/ui/src/views/jobs/index.tsx, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3drAvEKHZb7vyqGFQ:-Kq3n41yg2I6yAr4o5f9:bnwtyhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L75)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

We might not want to guess this, and instead just report time elapsed.  It will inevitably be unstable, and we've all been frustrated by progress bars with wildly inaccurate "time remaining" estimates.
</blockquote></details>

punting this question to @dianasaur and @kuanluo

---

*[pkg/ui/src/views/jobs/index.tsx, line 89 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3eS342qGM_eq5VQLe:-Kq3muGMw4V3QI8IRs-U:b-q0d38o) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L89)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Five digits seems like a ridiculous precision... is that likely to be useful?
</blockquote></details>

For large backup/restores it's nice to know if your job is hung or just very slow. Could probably knock it down to 3 though.

---

*[pkg/ui/src/views/jobs/index.tsx, line 99 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ZlQLNfMApYRRz7dt:-Kq3lE8dogMpY6uTnbkH:b-vvuup5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L99)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

nit: it's JSX that expects it to be capitalized (I suppose you could say TSX does as well, but only because it inherits the requirement).  AFAIC this comment is unnecessary, perhaps others disagree.
</blockquote></details>

https://github.com/cockroachdb/cockroach/blob/aa9112f8263033d62d790beb3faf95d3791186fd/pkg/ui/src/views/cluster/containers/events/index.tsx#L36-L41

(It's everywhere. I'm only a poor cargo-culter.)

---

*[pkg/ui/src/views/jobs/index.tsx, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3_NRNKs6lCeICa7Ig:-Kq3lQZwAxNQbxBsS4QE:bq8gp41) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L105)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Making this table sortable in combinations with the limit option is pretty awkward.  Usually the expectation of a sort on a table showing limited results is that it will do the same thing as the corresponding SQL query, with the ORDER BY applied _before_ the LIMIT.  Since the API endpoint doesn't accept a sort parameter, this won't work as expected.  I'd suggest either exposing the sort as an additional API parameter, or removing the sort option entirely.
</blockquote></details>

Yeah, I hear you, but I think this is still better than disabling sorting entirely. I agree it's weird UX, but we don't currently have a sortable table component that can forward sort requests to the backend, and there's no way that I'll have time to build that for 1.1.

---

*[pkg/ui/src/views/jobs/index.tsx, line 118 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3f9BO0GfWRlgDxV4C:-Kq3kNSCyUc1-Alp5yEP:b-8yl888) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L118)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

What's this for?
</blockquote></details>

The "JOBS" title at the top of the page. It's called by some layout somewhere.

---

*[pkg/ui/src/views/jobs/index.tsx, line 147 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ffy4gBjYlxEtNHXI:-Kq3lxHS_184l4A70YGi:b-bqixhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L147)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I think the ends of these tags should line up with their start (and we should have a lint rule for it...)
</blockquote></details>

I've got no opinion on the matter, but I think this is consistent with the way it's done elsewhere. Mind if I punt for now?

---

*[pkg/ui/styl/pages/jobs.styl, line 5 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3i_mbw56pccne1T2Q:-Kq3mOR6_Wu_Nx8hxeCP:bqd0tcy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/pages/jobs.styl#L5)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

What's this for?
</blockquote></details>

Tables are zebra-striped, providing a nice visual divide between two rows... except not when two failed rows appear next to each other.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-Kq3nM4UXrUHSWPT-dvq:b-1hvks8)*
<!-- Sent from Reviewable.io -->

kuanluo commented:
@benesch Cool, your plan sounds good to me.

danhhz commented:
> Thanks for all the reviews so far, everyone! Just a quick heads up that I'm swamped with other features for the 1.1 code freeze, so the more we can punt to August the better. I'm not sure how these things usually work, but in particular would love to land without full product/design QA: having this draft version merged will be hugely helpful as we debug backup/restore, and I'll have plenty of time for UX polishing in August.

Yeah, our bad. Should have been more clear. The intention here was to get _something_, even if not the final designs. Given @benesch's already insanely tight schedule for resumable backup/restore, I'm going to push back on doing any of the UX changes before August.

OTOH, my main concern was getting the server side changes in before code freeze. It may be worth splitting that part out into another PR and letting this stay open until everyone is happy. Given how much discussion has happened here already, it may not be worth cramming the UI side though

---

Review status: 16 of 21 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-Kq3p7sLE9tS7XO6V8bn:b-gv4lae)*
<!-- Sent from Reviewable.io -->

dianasaur323 commented:
Gotcha - Let's stop churning on this then. The UI looks good enough for now. @kuanluo, do you mind creating a new issue that includes your UI polishes so that we can track them there, and we can punt this until August.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-Kq3x3CCCKJtXiKDQTrX:b-mesrux)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 16 of 21 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 1019 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3K0mHAUMS4ZqYjFcy:-Kq41CvKWVC2jXGNcHhz:b-g3umoe) ([raw file](https://github.com/cockroachdb/cockroach/blob/fba0dd2edbec2a9468903b08ef5dfa174253d964/pkg/server/admin.go#L1019)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Just preferring public API to private. AIUI, @knz did a refactor so that `SHOW FOO` queries are identical to querying the underlying vtable via [`p.delegateQuery`](https://github.com/benesch/cockroach/blob/jobs-ui/pkg/sql/show.go#L572). So once we push filters down to the vtable constructors we should see that speedup in `SHOW JOBS` as well.
</blockquote></details>

1) in terms of efficiency this is already equivalent
2) yes I would like to suggest that `show` has the stable schemas whereas anything in `crdb_internal` is free-for-all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253)*
<!-- Sent from Reviewable.io -->

kuanluo commented:
Yup, done.  See #17257 for UI polishing.

vivekmenezes commented:
Incredible Nikhil!

On Thu, Jul 27, 2017 at 1:26 PM kuanluo <notifications@github.com> wrote:

> Yup, done. See #17257
> <https://github.com/cockroachdb/cockroach/issues/17257> for UI polishing.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/17253#issuecomment-318430321>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBPlh2z3p6h3_8V2ITBsc9DDcyxWBks5sSMhLgaJpZM4Ok6Pt>
> .
>

mrtracy commented:
Review status: 16 of 21 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[pkg/ui/src/views/shared/components/sortedtable/index.tsx, line 62 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq4-7QJkupaBVXlJb7d:-Kq4-7QJkupaBVXlJb7e:b-md3goh) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a609f2532cc8a174a9e4bd0fe32f2c31c56208/pkg/ui/src/views/shared/components/sortedtable/index.tsx#L62)):*
> ```XML
>  */
> export class SortedTable<T> extends React.Component<SortedTableProps<T>, {}> {
>     static defaultProps: Partial<SortedTableProps<any>> = {
> ```

Weird indentation here.

---

*[pkg/ui/styl/base/palette.styl, line 39 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ierGzLmVFh3Bk4Dr:-Kq437mIDQp4FkTNP8qP:bwhzt47) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/base/palette.styl#L39)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Looks like we're using upper-case hex codes.
</blockquote></details>

We should avoid adding another element-specific color here, it gets out of hand and you end up with 20 indistinguishably different shades of gray.

I would re-use $notification-alert-fill-color. Although that name may seem too specific, I don't think "notification" should be taken to mean literally notification popups, but rather anything attempting to get the users attention.   Perhaps with a rename (don't rename it in this PR) it would be more clear, but I think it's the correct color variable to use.

An alternative would be to use `lighten($alert-color, 90%)` wherever you are using this color.

@kuan 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253)*
<!-- Sent from Reviewable.io -->

benesch commented:
Yeah, @mrtracy, $notification-alert-fill-color is way too bright. Your
lighten solution SGTM, though.

On Thu, Jul 27, 2017 at 1:36 PM, Matt Tracy <notifications@github.com>
wrote:

> Review status: 16 of 21 files reviewed at latest revision, 21 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *pkg/ui/src/views/shared/components/sortedtable/index.tsx, line 62 at r3
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq4-7QJkupaBVXlJb7d:-Kq4-7QJkupaBVXlJb7e:b-md3goh>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/19a609f2532cc8a174a9e4bd0fe32f2c31c56208/pkg/ui/src/views/shared/components/sortedtable/index.tsx#L62>):*
>
>  */
> export class SortedTable<T> extends React.Component<SortedTableProps<T>, {}> {
>     static defaultProps: Partial<SortedTableProps<any>> = {
>
> Weird indentation here.
> ------------------------------
>
> *pkg/ui/styl/base/palette.styl, line 39 at r2
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ierGzLmVFh3Bk4Dr:-Kq437mIDQp4FkTNP8qP:bwhzt47>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/base/palette.styl#L39>):*
> *Previously, couchand (Andrew Couch) wrote…*
>
> Looks like we're using upper-case hex codes.
>
> We should avoid adding another element-specific color here, it gets out of
> hand and you end up with 20 indistinguishably different shades of gray.
>
> I would re-use $notification-alert-fill-color. Although that name may
> seem too specific, I don't think "notification" should be taken to mean
> literally notification popups, but rather anything attempting to get the
> users attention. Perhaps with a rename (don't rename it in this PR) it
> would be more clear, but I think it's the correct color variable to use.
>
> An alternative would be to use lighten($alert-color, 90%) wherever you
> are using this color.
>
> @Kuan <https://github.com/kuan>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17253>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/17253#issuecomment-318432776>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15ILnJhVml7hIMdvesE48ximfmqqMEks5sSMqWgaJpZM4Ok6Pt>
> .
>

mrtracy commented:
A lot of suggestions about optimizations, but I don't think we have sufficient evidence that things are slow.  We haven't established optimization guidelines for our project, and it's going to be annoying to bring those up in PRs going forward, *especially* when the existing codebase hasn't been optimized in that way.

Since we are trying to get this done for 1.1 feature complete, I would suggest that most of the suggested optimization work can be left for later; either in the August time frame, or even perhaps to future releases.  There is definitely an opportunity to go over the entire UI code base and apply a number of these optimizations (e.g. moving instantiations out of render methods where possible), but that will not preclude our 1.1 release.

Even if the jobs table is checked in and is very slow for these reasons, performance tuning would fall into the bug-related work slated for August.

---

Review status: 16 of 21 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[pkg/ui/src/views/jobs/index.tsx, line 15 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3YnKYa391cxjD1jPu:-Kq4FyeT7bONGo4J6i2h:b-z30tk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L15)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Damn, I spent like 10m trying to make sense of the ordering rules last night. SGTM.
</blockquote></details>

You were attempting to make sense of a dumpster fir- I mean, a modern art installation. Sorry

---

*[pkg/ui/src/views/jobs/index.tsx, line 91 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ee-XN6xmKDhsZacu:-Kq4GjY3Kosr6FTGMZrm:b-91a9vc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L91)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

This might be better done with explicit styling.
</blockquote></details>

I'm on the fence here, I think `em` is semantically fine, but indeed we really don't know *what* styling will be applied to this element, we're depending on a sane default from the browser or from a setting somewhere else.  Being explicit is probably better.

---

*[pkg/ui/src/views/jobs/index.tsx, line 99 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ZlQLNfMApYRRz7dt:-Kq4FbdTvNN0YTWq7VVV:bxnu42v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L99)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/blob/aa9112f8263033d62d790beb3faf95d3791186fd/pkg/ui/src/views/cluster/containers/events/index.tsx#L36-L41

(It's everywhere. I'm only a poor cargo-culter.)
</blockquote></details>

I feel that almost any lint rule override deserves a comment.

---

*[pkg/ui/src/views/jobs/index.tsx, line 147 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ffy4gBjYlxEtNHXI:-Kq49TWnE6UZDNWid0ej:bkxqmmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L147)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I've got no opinion on the matter, but I think this is consistent with the way it's done elsewhere. Mind if I punt for now?
</blockquote></details>

We're going to install a linter rule for it later, this will just get caught up in the sweep.  If you have other round-trip changes to make go ahead and put it right, but otherwise yeah it's fine to punt.

---

*[pkg/ui/src/views/jobs/index.tsx, line 188 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3hIveBTdoFtxiilsW:-Kq4AdzXyt4bXt95wo2V:b-34lfbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L188)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

We should try to minimize calls to the `moment` constructor.  It's fairly expensive, and we're just repeating work if we do it on every render.  It's much better to do that once when we get the API results than repeatedly in component renders.  Moving it there might be a significant change, since the `timeoutFetch` and `CachedDataReducer` abstract away all of that side of things.  But we can still improve somewhat by moving it to the `mapStateToProps` function and only do it once per state change.
</blockquote></details>

It's not readily apparent to me that this would be the correct optimization. Even assuming that TimestampToMoment is expensive and should be memoized, converting timestamps to moments at the API level is almost a non-starter, as those object definitions are generated from protocol buffers. To do it here at the table component, we would have to run the 'jobs' through a selector which maps them into objects with shape `{ originalJob, startTimestamp }`, which would definitely make the code a bit more difficult to understand.

And in either of those cases, you'd be running TimestampToMoment on every job every time data is fetched; if the number of rows being fetched is very large, you would expect to render a limited set of the original data anyway, in which case calling TimestampToMoment at render time is the more efficient choice.  If the number of rows being fetched is *not* large (and here it's explicitly limited), I would be surprised if memoizing would make a difference worth the extra development overhead, since it would only benefit some types of renders.

Nobody has yet taken the time to performance-profile the UI in depth, and I think we should be careful suggesting optimizations in PRs unless they are exceedingly straightforward.

---

*[pkg/ui/src/views/jobs/index.tsx, line 211 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3gw3PQsPdZLPDiU6C:-Kq47S543lDKHv2lH8xT:bcb7ga6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L211)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I prefer to extract the `mapStateToProps` and `mapDispatchToProps` (or in this case, just `actions`) so that the `connect` call can be one-lined.
</blockquote></details>

After having it written in this inline style everywhere, and then seeing it written with separate functions, I'm going to agree with Andrew. "connect" returns a function which is then invoked with the unconnected class, and the return from that function is exported; this is complicated behavior that is more clear in the one-line form.

---

*[pkg/ui/styl/pages/jobs.styl, line 5 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3i_mbw56pccne1T2Q:-Kq46sFb_YiRkVnz9l_s:b-tsclxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/pages/jobs.styl#L5)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Tables are zebra-striped, providing a nice visual divide between two rows... except not when two failed rows appear next to each other.
</blockquote></details>

Add a `//` comment, the syntax is fine but once you get enough ampersands it becomes a bit inscrutable.

---

*[pkg/ui/styl/pages/jobs.styl, line 13 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3iwwmuXzmUmAy2SDl:-Kq474m7I-J4PCUMM4TB:b-74xe53) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/pages/jobs.styl#L13)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

This should have a fallback generic family.
</blockquote></details>

We are guilty of this everywhere, it would likely be best to let this PR go ahead and log a bug to deal with font fallbacks everywhere in our stylus code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-Kq4HDid3n_ZEWE4pQs3:b-r0ctmp)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 6 of 25 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[pkg/server/admin.go, line 1340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3VgffvWBxJmHOwRpQ:-Kq53DNrxbpDFSpjGQDw:bhiqko1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/server/admin.go#L1340)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

nit: "only" will seem odd if too many dests are provided
</blockquote></details>

Done in #17260. Thanks!

---

*[pkg/ui/src/views/jobs/index.tsx, line 15 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3YnKYa391cxjD1jPu:-Kq557AerIhtTpJVg86s:bscbkoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L15)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

You were attempting to make sense of a dumpster fir- I mean, a modern art installation. Sorry
</blockquote></details>

Heh, done.

---

*[pkg/ui/src/views/jobs/index.tsx, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3drAvEKHZb7vyqGFQ:-Kq559AkWQHS4JQE14BU:blhbwxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L75)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

punting this question to @dianasaur and @kuanluo
</blockquote></details>

Ok, going to leave this in for now.

---

*[pkg/ui/src/views/jobs/index.tsx, line 89 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3eS342qGM_eq5VQLe:-Kq55Crr4Y_fXSxckmvS:b8hzckh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L89)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

For large backup/restores it's nice to know if your job is hung or just very slow. Could probably knock it down to 3 though.
</blockquote></details>

Turned down to 3.

---

*[pkg/ui/src/views/jobs/index.tsx, line 91 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ee-XN6xmKDhsZacu:-Kq5JNuoEWgJ4vsrhcZ0:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L91)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I'm on the fence here, I think `em` is semantically fine, but indeed we really don't know *what* styling will be applied to this element, we're depending on a sane default from the browser or from a setting somewhere else.  Being explicit is probably better.
</blockquote></details>

Fixed.

---

*[pkg/ui/src/views/jobs/index.tsx, line 99 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ZlQLNfMApYRRz7dt:-Kq56kxky2C1SSc2vMqj:b-3tam3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L99)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I feel that almost any lint rule override deserves a comment.
</blockquote></details>

Ok, updated everywhere to say "JSX" instead of "TSX".

---

*[pkg/ui/src/views/jobs/index.tsx, line 146 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ftzOqeNvWIFVMH54:-Kq57JR2EbISeaFowar0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L146)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Inlining handlers is not ideal since it prevents a component from shallowly checking its props for changes.  It's better to make a function elsewhere (usually as a method of this component) that will be identical across rerenders.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/jobs/index.tsx, line 147 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ffy4gBjYlxEtNHXI:-Kq56zEKKkgxqctxs38E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L147)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

We're going to install a linter rule for it later, this will just get caught up in the sweep.  If you have other round-trip changes to make go ahead and put it right, but otherwise yeah it's fine to punt.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/jobs/index.tsx, line 174 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3gjTZQuPU9jumutoY:-Kq57KYabgxJ9ItZfL6H:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L174)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

This column configuration should be defined above to avoid allocations and so that it's identical across rerenders.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/jobs/index.tsx, line 188 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3hIveBTdoFtxiilsW:-Kq57Lp_qcx9v22PYwEv:b-afpdud) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L188)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

It's not readily apparent to me that this would be the correct optimization. Even assuming that TimestampToMoment is expensive and should be memoized, converting timestamps to moments at the API level is almost a non-starter, as those object definitions are generated from protocol buffers. To do it here at the table component, we would have to run the 'jobs' through a selector which maps them into objects with shape `{ originalJob, startTimestamp }`, which would definitely make the code a bit more difficult to understand.

And in either of those cases, you'd be running TimestampToMoment on every job every time data is fetched; if the number of rows being fetched is very large, you would expect to render a limited set of the original data anyway, in which case calling TimestampToMoment at render time is the more efficient choice.  If the number of rows being fetched is *not* large (and here it's explicitly limited), I would be surprised if memoizing would make a difference worth the extra development overhead, since it would only benefit some types of renders.

Nobody has yet taken the time to performance-profile the UI in depth, and I think we should be careful suggesting optimizations in PRs unless they are exceedingly straightforward.
</blockquote></details>

Ok, leaving this be for now.

---

*[pkg/ui/src/views/jobs/index.tsx, line 211 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3gw3PQsPdZLPDiU6C:-Kq57iUqIa3cM7_2oRWa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L211)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

After having it written in this inline style everywhere, and then seeing it written with separate functions, I'm going to agree with Andrew. "connect" returns a function which is then invoked with the unconnected class, and the return from that function is exported; this is complicated behavior that is more clear in the one-line form.
</blockquote></details>

Done.

---

*[pkg/ui/src/views/shared/components/sortedtable/index.tsx, line 62 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq4-7QJkupaBVXlJb7d:-Kq57ocXa6X6eMJb7ZEz:brfv4j8) ([raw file](https://github.com/cockroachdb/cockroach/blob/19a609f2532cc8a174a9e4bd0fe32f2c31c56208/pkg/ui/src/views/shared/components/sortedtable/index.tsx#L62)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Weird indentation here.
</blockquote></details>

Thanks. I'm used to having `gofmt` around.

---

*[pkg/ui/styl/pages/jobs.styl, line 5 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3i_mbw56pccne1T2Q:-Kq58Dpi52yxgDTJrOHm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/pages/jobs.styl#L5)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Add a `//` comment, the syntax is fine but once you get enough ampersands it becomes a bit inscrutable.
</blockquote></details>

Done.

---

*[pkg/ui/styl/pages/jobs.styl, line 13 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3iwwmuXzmUmAy2SDl:-Kq58MRL5AfeycmNN8Mp:bkdfkno) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/pages/jobs.styl#L13)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

We are guilty of this everywhere, it would likely be best to let this PR go ahead and log a bug to deal with font fallbacks everywhere in our stylus code.
</blockquote></details>

Ok, leaving alone.

---

*[pkg/ui/styl/pages/jobs.styl, line 14 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ikwUe93Wc1NibJlX:-Kq58OU-QJND1iF05cIB:bo3vl6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/pages/jobs.styl#L14)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Trailing whitespace.
</blockquote></details>

Thanks, done.

---

*[pkg/ui/styl/base/palette.styl, line 39 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3ierGzLmVFh3Bk4Dr:-Kq57xE69tyj4QgVcaaK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/styl/base/palette.styl#L39)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

We should avoid adding another element-specific color here, it gets out of hand and you end up with 20 indistinguishably different shades of gray.

I would re-use $notification-alert-fill-color. Although that name may seem too specific, I don't think "notification" should be taken to mean literally notification popups, but rather anything attempting to get the users attention.   Perhaps with a rename (don't rename it in this PR) it would be more clear, but I think it's the correct color variable to use.

An alternative would be to use `lighten($alert-color, 90%)` wherever you are using this color.

@kuan 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253)*
<!-- Sent from Reviewable.io -->

benesch commented:
Also, @kuanluo, do you have an icon for this page?

kuanluo commented:
@benesch Would these represent jobs? Quite literally...

Active state | Inactive state
------------ | -------------
![briefcase-24 2x](https://user-images.githubusercontent.com/1486483/28782687-e8b2139e-75db-11e7-973b-2711f8310e72.png) | ![briefcase-24 2x 1](https://user-images.githubusercontent.com/1486483/28782688-e8b99452-75db-11e7-93dc-33eeb5403756.png)

benesch commented:
Hah, love it. Looks good to me.

On Mon, Jul 31, 2017 at 10:38 AM, kuanluo <notifications@github.com> wrote:

> @benesch <https://github.com/benesch> Would these represent jobs? Quite
> literally...
> Active state Inactive state
> [image: briefcase-24 2x]
> <https://user-images.githubusercontent.com/1486483/28782687-e8b2139e-75db-11e7-973b-2711f8310e72.png> [image:
> briefcase-24 2x 1]
> <https://user-images.githubusercontent.com/1486483/28782688-e8b99452-75db-11e7-93dc-33eeb5403756.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/17253#issuecomment-319086516>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IOwc54fhbUIIrcMnJfWtrvZpryyHks5sTebYgaJpZM4Ok6Pt>
> .
>

benesch commented:
Ok, PTAL! I've updated to use @kuanluo's new icon, plus I think I've addressed all review feedback so far.

benesch commented:
Ping @mrtracy @couchand.

couchand commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice work!

---

Reviewed 16 of 27 files at r5.
Review status: 16 of 20 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[pkg/ui/src/views/jobs/index.tsx, line 75 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3drAvEKHZb7vyqGFQ:-KqxSN02Y-ucFBcFRVEs:b76f16w) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L75)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ok, going to leave this in for now.
</blockquote></details>

Seems fine.

---

*[pkg/ui/src/views/jobs/index.tsx, line 105 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3_NRNKs6lCeICa7Ig:-KqxSea4Rix_1kGdrWMl:bw63w0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L105)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Yeah, I hear you, but I think this is still better than disabling sorting entirely. I agree it's weird UX, but we don't currently have a sortable table component that can forward sort requests to the backend, and there's no way that I'll have time to build that for 1.1.
</blockquote></details>

Fair point.

---

*[pkg/ui/src/views/jobs/index.tsx, line 188 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3hIveBTdoFtxiilsW:-KqxT-Ga2lHIS1svCuzU:bq8rxiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L188)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ok, leaving this be for now.
</blockquote></details>

Fine to leave this for now.  But I maintain (from past experience profiling UIs that use it) that `moment` is way too expensive for general use and should be burninated in the 1.2 timeline.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-KqxUdpLwGqfahTtgkje:b-kbynj8)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 16 of 20 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-KqyQDIYwfcunGrtlurS:bnfp4nl)*
<!-- Sent from Reviewable.io -->

benesch commented:
TFTRs!

---

Review status: 16 of 20 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[pkg/ui/src/views/jobs/index.tsx, line 188 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3hIveBTdoFtxiilsW:-KqySkDWHVIC0XVK7inr:b-eetxpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bce04ab73ed36d06b2af93ee6f4a150bab04a53/pkg/ui/src/views/jobs/index.tsx#L188)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Fine to leave this for now.  But I maintain (from past experience profiling UIs that use it) that `moment` is way too expensive for general use and should be burninated in the 1.2 timeline.
</blockquote></details>

Yeah, I hear you. Happy to help in that effort! I have no great love for moment.

---

*[pkg/server/admin.go, line 1019 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-Kq3K0mHAUMS4ZqYjFcy:-KqZqe-IxhyrOlYdLMhT:b-gjr387) ([raw file](https://github.com/cockroachdb/cockroach/blob/fba0dd2edbec2a9468903b08ef5dfa174253d964/pkg/server/admin.go#L1019)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) in terms of efficiency this is already equivalent
2) yes I would like to suggest that `show` has the stable schemas whereas anything in `crdb_internal` is free-for-all.
</blockquote></details>

Done in #17260.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17253#-:-KqySg7dOVwmZW9Vdbwy:bz5mair)*
<!-- Sent from Reviewable.io -->

benesch committed:
Merge pull request #17409 from benesch/scatter-resp

storage: restore range information in AdminScatterResponse
